lib/changesreader.js

```diff
@@ -0,0 +1,284 @@
+    this.continue = true // whether to poll again
+    this.qs = {} // extra querystring parameters
+    this.selector = null
+  }
+
+  // prevent another poll happening
+  stop () {
+    this.continue = false
+  }
+
+  // sleep, promise style
+  async sleep (t) {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => { resolve() }, t)
+    })
+  }
+
+  // called to start listening to the changes feed. The opts object can contain:
+  // - batchSize - the number of records to return per HTTP request
+  // - since - the the sequence token to start from (defaults to 'now')
+  start (opts) {
+    const self = this
+
+    // if we're already listening for changes
+    if (self.started) {
+      // return the existing event emitter
+      return self.ee
+    }
+    self.started = true
+
+    // handle overidden defaults
+    opts = opts || {}
+    Object.assign(self, opts)
+
+    // the work function is async and runs in the background
+    // with a big do/while loop
+    const work = async () => {
+      do {
+        // formulate changes feed longpoll HTTP request
+        let pause = 0
+        const req = {
+          method: 'post',
+          db: self.db,
+          path: '_changes',
+          qs: {
+            feed: 'longpoll',
+            timeout: self.timeout,
+            since: self.since,
+            limit: self.batchSize,
+            include_docs: self.includeDocs
+          },
+          body: {}
+        }
+        if (self.fastChanges) {
+          req.qs.seq_interval = self.batchSize
+        }
+        if (self.selector) {
+          req.qs.filter = '_selector'
+          req.body.selector = self.selector
+        }
+        Object.assign(req.qs, opts.qs)
+
+        // make HTTP request to get up to batchSize changes from the feed
+        try {
+          const data = await self.request(req)
+
+          // update the since state
+          if (data && data.last_seq && data.last_seq !== self.since) {
+            self.since = data.last_seq
+            self.ee.emit(EVENT_SEQ, self.since)
+          }
+
+          // stop on empty batch or small batch
+          if (self.stopOnEmptyChanges && data && typeof data.results !== 'undefined' && data.results.length < self.batchSize) {
+            self.continue = false
+          }
+
+          // if we have data
+          if (data && data.results && data.results.length > 0) {
+            // emit EVENT_CHANGE events
+            for (const i in data.results) {
+              self.ee.emit(EVENT_CHANGE, data.results[i])
+            }
+
+            // batch event
+            // emit EVENT_BATCH event
+            if (self.wait) {
+              // in 'wait' mode, we need to wait until the user calls
+              // a 'done' function before issuing the next change request
+              await new Promise((resolve, reject) => {
+                self.ee.emit(EVENT_BATCH, data.results, () => {
+                  resolve()
+                })
+              })
```

Comment:
This section was quite surprising to me.

Listeners are typically used for receiving updates on what is happening under the hood, and in this case this is being used to control the execution.

> Also no need to create a lambda here, simply use `resolve`

Reply:
It is a little unusual, and is certainly optional behaviour. The only other way I can think of doing it is to provide another method (`resume`, say) that the user has to remember to call after each batch of changes.

Agreed on the `resolve` bit

Reply:
Yes, that approach seems more idiomatic

Also, in order to get into wait-mode the user had to set the flag, so I think it kind of makes sense to require making such call before continuing reading changes.

Additionally this approach could help split reading changes from the server from the actual code processing those changes, and so by having two different executions contexts could provide a speed boost as one context is always reading without needing to wait for those to be processed.

The only downside is that the change reader could put pressure on slow change processors, but there could be a configurable threshold for how many changes can be queued while waiting for the slow processors.

Please let me know your thoughts about this idea, I might be missing something in here

Reply:
I'm going to see about add pause/resume methods to avoid the callback-in-the-event scenario